### ISSUE TITLE
Renderers fully replace DOM, losing state such as focus

### ISSUE THREAD

**Issue by sissbruecker** [vaadin/react-components#283]
Currently all renderers (grid cells, dashboard widgets, ...) are treated as actual components rather than just functions that render children. That means that the React virtual DOM will contain an instance of that component, rather than just the children that the function returns. That is problematic with how we advertise using renderers in the docs:
```tsx
<Grid items={items.value}>
  <GridColumn header="Employee">
    {({ item }) => <span class="foo">{item.fullName}</span>}
  </GridColumn>

  ...
</Grid>
```

- With this example, it creates a new function on each render cycle
- Since we treat those as a component, it means it actually creates a new component type each render cycle
- So the virtual DOM contains a different type of component on each render cycle
- **That results in React fully replacing the actual DOM, and any state is lost (focus, text selection, entered text)**

From just looking at the code, this is also easy to miss:
- It's not obvious that the function is now an actual component and that it changes how React updates the DOM
- A render function is a common concept in the React ecosystem and the example above looks like one. However a render function is typically just called directly instead of creating a component instance from it

### Option 1: Treat renderers as just render functions

We should consider treating renderers as as just render functions in their traditional sense. So if the renderer is a function, we just call that function instead of creating a component instance with `React.createElement`.

By doing that, the virtual DOM will always contain the same type of component. For example, if a function returns a button, then the virtual DOM will always just contain a button. That allows React to update the current DOM instead of replacing it completely.

We need to check if this can be done without breaking changes:
- We currently allow passing any type of React component, included class-based components, which can not be called as a function. As a workaround, we might be able to render as component as we do now, but then discard it and instead add its children to the virtual DOM.
- Long term, we should only allow passing functions.
- Need to check if there is some way to use refs with renderers that could break.

### Option 2: Improve docs to clarify that renderers are components

If we actually want renderers to be components, we could update our docs make sure that users are guided to never recreate renderers on each render cycle. For example by memoizing the component like so:
```tsx
const employeeRenderer = useCallback(
  ({ item }) => <span class="foo">{item.fullName}</span>,
  []
);
    
<Grid items={items.value}>
  <GridColumn header="Employee">
    {employeeRenderer}
  </GridColumn>

  ...
</Grid>
```

However, using components make trivial use-cases more complex. For example:
```tsx
const [selectedItems, setSelectedItems] = useState<Person[]>();
	
const selectionRenderer = useCallback(
  ({ item }) => (selectedItems.includes(item) ? <SelectedIcon> : <UnselectedIcon>),
  []
);
    
<Grid items={items.value}>
  <GridColumn>
    {selectionRenderer}
  </GridColumn>

  ...
</Grid>
```

Here a renderer uses the selection state from the component that it's used in. The problem is that it doesn't actually work, because the renderer component will always reference the initial selection state. To fix this, a new renderer component would have to be created each time the selection changes, which brings us back to the initial problem. There are ways to solve this using React contexts or signals, but still:
- The complexity increases
- There is potential for subtle bugs

**Comment by yuriy-fix** [comment]
Let's invest 1-2 days into researching the option 1 and if it would cause breaking changes.

**Comment by vursen** [comment]
I'm not sure why class-based components are a concern for option 1. Classes are typically declared outside of the React render scope, so their reference should remain stable across renders, shouldn't it?

**Comment by vursen** [comment]
Simply calling the function turns out to be unsafe, as that function might use hooks, which are only allowed inside components. There is a warning that React displays if the function uses hooks and is called directly:

![image](https://github.com/user-attachments/assets/e3ad1040-d00f-4a82-8d0d-dd47c4e8e1bf)

And at least AutoGrid uses hooks widely in [its renderer functions](https://github.com/vaadin/hilla/blob/67be5edcbc3a57f1e191b2c12c5b8b04eeb68484/packages/ts/react-crud/src/autogrid-renderers.tsx#L54-L58).

**Comment by sissbruecker** [comment]
Good point, treating components as functions indeed doesn't work with hooks. Regarding the extracting children approach, I was assuming that it would be possible to extract the children from a React component rendered with `createElement`, however it appears that is not possible.

The approach outlined above seems like a reasonable compromise though having to wrap a render function is indeed odd and not very discoverable. An alternative could be to add a new API where render functions must always be passed as a separate prop (not through children), and which use a different naming convention than the current renderer props (e.g. `renderSomething` instead of `somethingRenderer`). Then deprecate the current API, log warnings around its usage and update docs examples. Once the old API is eventually removed we could restore passing a render function as child. But that would result in a lot of effort for users and ourselves.

**Comment by vursen** [comment]
> The approach outlined above seems like a reasonable compromise though having to wrap a render function is indeed odd and not very discoverable

During the transition phase, we could display a warning that component functions are deprecated, and in the next major, we might throw an exception to enforce the wrapper usage.

> An alternative could be to add a new API where render functions must always be passed as a separate prop

That's also an option 👍

**Comment by yuriy-fix** [comment]
We should probably proceed with updating documentation examples for now.

**Comment by sissbruecker** [comment]
Extracted https://github.com/vaadin/react-components/issues/294 for updating the documentation. This issue is now about fixing the root cause of not treating renderers as components.

**Comment by sissbruecker** [comment]
Another issue that came up is that this can prevent click listeners from working if a blur event triggers a state change in the component, causing the clicked element to be recreated. The flow of events is like this:
- Focus a text field
- Mousedown on a button created by a renderer
- Triggers blur listener on the text field, listener changes React state
- React component rerenders, new instance of the renderer function causes the old button to be replaced with a new one
- Mouseup doesn't trigger a click event, because the new button never received a mousedown event

A sample reproduction of this using a dialog and Hilla's `useForm` hook:
```tsx
export default function DemoView() {
  const opened = useSignal(false);
  const form = useForm(PersonModel);

  return (
    <>
      <Dialog
        headerTitle="Add user"
        opened={opened.value}
        headerRenderer={() => (
          <Button theme="tertiary" onClick={() => (opened.value = false)}>Close</Button>
        )}
        footerRenderer={() => (
          <Button disabled={form.invalid} onClick={() => (opened.value = false)}>
            Save
          </Button>
        )}>
        <TextField label="Name" {...form.field(form.model.name)} />
      </Dialog>
      <Button onClick={() => (opened.value = true)}>Open dialog</Button>
    </>
  );
}
```

In this example first focusing the field, and then clicking the close button in the header does nothing. On blur the `useForm` hook changes internal state causing the dialog to rerender. As part of that a new close button is rendered, and the click event is never triggered.

To fix this one needs to extract the header and footer renderers and memoize them so that they are only created once over the component's lifetime. Since the footer renderer uses the invalid state of the form, the invalid state needs to be extracted into a signal, otherwise the renderer would otherwise always reference the initial state of the form.

```tsx
  const opened = useSignal(false);
  const form = useForm(PersonModel);
  const invalid = useSignal(false);
  invalid.value = form.invalid;

  const headerRenderer = useCallback(
    () => (
      <Button theme="tertiary" onClick={() => (opened.value = false)}>
        Close
      </Button>
    ),
    [],
  );

  const footerRenderer = useCallback(
    () => (
      <Button disabled={invalid.value} onClick={() => (opened.value = false)}>
        Save
      </Button>
    ),
    [],
  );

  return (
    <>
      <Dialog
        headerTitle="Add user"
        opened={opened.value}
        headerRenderer={headerRenderer}
        footerRenderer={footerRenderer}>
        <TextField label="Name" {...form.field(form.model.name)} />
      </Dialog>
      <Button onClick={() => (opened.value = true)}>Open dialog</Button>
    </>
  );
```
